File: kernel/src/mod/rust/drm/resources.c
#include <mod/rust/drm/internal.h>

void na_drm_mode_copy(struct drm_mode_modeinfo *destination,
                      const na_drm_mode_info_t *source) {
    *destination = (struct drm_mode_modeinfo){
        .clock = source->clock,
        .hdisplay = source->hdisplay,
        .hsync_start = source->hsync_start,
        .hsync_end = source->hsync_end,
        .htotal = source->htotal,
        .hskew = source->hskew,
        .vdisplay = source->vdisplay,
        .vsync_start = source->vsync_start,
        .vsync_end = source->vsync_end,
        .vtotal = source->vtotal,
        .vscan = source->vscan,
        .vrefresh = source->vrefresh,
        .flags = source->flags,
        .type = source->mode_type,
    };
    memcpy(destination->name, source->name, sizeof(destination->name));
}

void na_drm_mode_export(na_drm_mode_info_t *destination,
                        const struct drm_mode_modeinfo *source) {
    *destination = (na_drm_mode_info_t){
        .clock = source->clock,
        .hdisplay = source->hdisplay,
        .hsync_start = source->hsync_start,
        .hsync_end = source->hsync_end,
        .htotal = source->htotal,
        .hskew = source->hskew,
        .vdisplay = source->vdisplay,
        .vsync_start = source->vsync_start,
        .vsync_end = source->vsync_end,
        .vtotal = source->vtotal,
        .vscan = source->vscan,
        .vrefresh = source->vrefresh,
        .flags = source->flags,
        .mode_type = source->type,
    };
    memcpy(destination->name, source->name, sizeof(destination->name));
}

static uint32_t na_drm_resource_token(uint32_t index, uint32_t capacity) {
    return index < capacity ? index + 1 : 0;
}

drm_connector_t *na_drm_connector_create(const na_drm_connector_info_t *info) {
    if (!info ||
        (info->encoder_index != UINT32_MAX &&
         info->encoder_index >= DRM_MAX_ENCODERS_PER_DEVICE) ||
        (info->crtc_index != UINT32_MAX &&
         info->crtc_index >= DRM_MAX_CRTCS_PER_DEVICE))
        return NULL;

    drm_connector_t *connector = calloc(1, sizeof(*connector));
    if (!connector)
        return NULL;

    connector->type = info->connector_type;
    connector->connection = info->connection;
    connector->encoder_id =
        na_drm_resource_token(info->encoder_index, DRM_MAX_ENCODERS_PER_DEVICE);
    connector->crtc_id =
        na_drm_resource_token(info->crtc_index, DRM_MAX_CRTCS_PER_DEVICE);
    connector->mm_width = info->mm_width;
    connector->mm_height = info->mm_height;
    connector->subpixel = info->subpixel;
    connector->refcount = 1;
    return connector;
}

int na_drm_connector_add_mode(drm_connector_t *connector,
                              const na_drm_mode_info_t *mode) {
    if (!connector || !mode || connector->count_modes == UINT32_MAX)
        return -EINVAL;

    size_t count = (size_t)connector->count_modes + 1;
    if (count > SIZE_MAX / sizeof(*connector->modes))
        return -EINVAL;
    struct drm_mode_modeinfo *modes =
        realloc(connector->modes, count * sizeof(*modes));
    if (!modes)
        return -ENOMEM;

    connector->modes = modes;
    na_drm_mode_copy(&modes[connector->count_modes], mode);
    connector->count_modes++;
    return 0;
}

void na_drm_connector_destroy(drm_connector_t *connector) {
    if (!connector)
        return;
    free(connector->modes);
    free(connector);
}

drm_crtc_t *na_drm_crtc_create(const na_drm_crtc_info_t *info) {
    if (!info)
        return NULL;

    drm_crtc_t *crtc = calloc(1, sizeof(*crtc));
    if (!crtc)
        return NULL;

    crtc->x = info->x;
    crtc->y = info->y;
    crtc->w = info->width;
    crtc->h = info->height;
    crtc->gamma_size = info->gamma_size;
    crtc->mode_valid = info->mode_valid;
    if (info->mode_valid)
        na_drm_mode_copy(&crtc->mode, &info->mode_info);
    crtc->refcount = 1;
    return crtc;
}

void na_drm_crtc_destroy(drm_crtc_t *crtc) { free(crtc); }

drm_encoder_t *na_drm_encoder_create(const na_drm_encoder_info_t *info) {
    if (!info || (info->crtc_index != UINT32_MAX &&
                  info->crtc_index >= DRM_MAX_CRTCS_PER_DEVICE))
        return NULL;

    drm_encoder_t *encoder = calloc(1, sizeof(*encoder));
    if (!encoder)
        return NULL;

    encoder->type = info->encoder_type;
    encoder->crtc_id =
        na_drm_resource_token(info->crtc_index, DRM_MAX_CRTCS_PER_DEVICE);
    encoder->possible_crtcs = info->possible_crtcs;
    encoder->possible_clones = info->possible_clones;
    encoder->refcount = 1;
    return encoder;
}

void na_drm_encoder_destroy(drm_encoder_t *encoder) { free(encoder); }

drm_plane_t *na_drm_plane_create(drm_device_t *device,
                                 const na_drm_plane_info_t *info) {
    if (!device || !info ||
        (info->crtc_index != UINT32_MAX &&
         info->crtc_index >= DRM_MAX_CRTCS_PER_DEVICE))
        return NULL;

    drm_plane_t *plane = calloc(1, sizeof(*plane));
    if (!plane)
        return NULL;

    if (info->crtc_index != UINT32_MAX &&
        device->resource_mgr.crtcs[info->crtc_index])
        plane->crtc_id = device->resource_mgr.crtcs[info->crtc_index]->id;
    plane->possible_crtcs = info->possible_crtcs;
    plane->gamma_size = info->gamma_size;
    plane->plane_type = info->plane_type;
    plane->refcount = 1;
    return plane;
}

int na_drm_plane_add_format(drm_plane_t *plane, uint32_t format) {
    if (!plane || plane->count_format_types == UINT32_MAX)
        return -EINVAL;

    size_t count = (size_t)plane->count_format_types + 1;
    if (count > SIZE_MAX / sizeof(*plane->format_types))
        return -EINVAL;
    uint32_t *formats = realloc(plane->format_types, count * sizeof(*formats));
    if (!formats)
        return -ENOMEM;

    plane->format_types = formats;
    formats[plane->count_format_types++] = format;
    return 0;
}

void na_drm_plane_destroy(drm_plane_t *plane) {
    if (!plane)
        return;
    free(plane->format_types);
    free(plane);
}

int na_drm_get_connectors(drm_device_t *device, drm_connector_t **connectors,
                          uint32_t *count) {
    na_drm_driver_bridge_t *bridge = na_drm_bridge(device);
    if (!bridge || !connectors || !count || !bridge->ops.get_connectors)
        return -ENOSYS;
    int result =
        bridge->ops.get_connectors(bridge->ops.context, device, connectors,
                                   DRM_MAX_CONNECTORS_PER_DEVICE, count);
    if (result < 0 || *count <= DRM_MAX_CONNECTORS_PER_DEVICE)
        return result;
    for (uint32_t i = 0; i < DRM_MAX_CONNECTORS_PER_DEVICE; i++)
        na_drm_connector_destroy(connectors[i]);
    *count = 0;
    return -EINVAL;
}

int na_drm_get_crtcs(drm_device_t *device, drm_crtc_t **crtcs,
                     uint32_t *count) {
    na_drm_driver_bridge_t *bridge = na_drm_bridge(device);
    if (!bridge || !crtcs || !count || !bridge->ops.get_crtcs)
        return -ENOSYS;
    int result = bridge->ops.get_crtcs(bridge->ops.context, device, crtcs,
                                       DRM_MAX_CRTCS_PER_DEVICE, count);
    if (result < 0 || *count <= DRM_MAX_CRTCS_PER_DEVICE)
        return result;
    for (uint32_t i = 0; i < DRM_MAX_CRTCS_PER_DEVICE; i++)
        na_drm_crtc_destroy(crtcs[i]);
    *count = 0;
    return -EINVAL;
}

int na_drm_get_encoders(drm_device_t *device, drm_encoder_t **encoders,
                        uint32_t *count) {
    na_drm_driver_bridge_t *bridge = na_drm_bridge(device);
    if (!bridge || !encoders || !count || !bridge->ops.get_encoders)
        return -ENOSYS;
    int result = bridge->ops.get_encoders(bridge->ops.context, device, encoders,
                                          DRM_MAX_ENCODERS_PER_DEVICE, count);
    if (result < 0 || *count <= DRM_MAX_ENCODERS_PER_DEVICE)
        return result;
    for (uint32_t i = 0; i < DRM_MAX_ENCODERS_PER_DEVICE; i++)
        na_drm_encoder_destroy(encoders[i]);
    *count = 0;
    return -EINVAL;
}

static void na_drm_finalize_links(drm_device_t *device) {
    for (uint32_t i = 0; i < DRM_MAX_CONNECTORS_PER_DEVICE; i++) {
        drm_connector_t *connector = device->resource_mgr.connectors[i];
        if (!connector)
            continue;

        uint32_t encoder_index = connector->encoder_id;
        connector->encoder_id =
            encoder_index && encoder_index <= DRM_MAX_ENCODERS_PER_DEVICE &&
                    device->resource_mgr.encoders[encoder_index - 1]
                ? device->resource_mgr.encoders[encoder_index - 1]->id
                : 0;

        uint32_t crtc_index = connector->crtc_id;
        connector->crtc_id =
            crtc_index && crtc_index <= DRM_MAX_CRTCS_PER_DEVICE &&
                    device->resource_mgr.crtcs[crtc_index - 1]
                ? device->resource_mgr.crtcs[crtc_index - 1]->id
                : 0;
    }

    for (uint32_t i = 0; i < DRM_MAX_ENCODERS_PER_DEVICE; i++) {
        drm_encoder_t *encoder = device->resource_mgr.encoders[i];
        if (!encoder)
            continue;
        uint32_t crtc_index = encoder->crtc_id;
        encoder->crtc_id = crtc_index &&
                                   crtc_index <= DRM_MAX_CRTCS_PER_DEVICE &&
                                   device->resource_mgr.crtcs[crtc_index - 1]
                               ? device->resource_mgr.crtcs[crtc_index - 1]->id
                               : 0;
    }
}

int na_drm_get_planes(drm_device_t *device, drm_plane_t **planes,
                      uint32_t *count) {
    na_drm_driver_bridge_t *bridge = na_drm_bridge(device);
    if (!bridge || !planes || !count || !bridge->ops.get_planes)
        return -ENOSYS;
    na_drm_finalize_links(device);
    int result = bridge->ops.get_planes(bridge->ops.context, device, planes,
                                        DRM_MAX_PLANES_PER_DEVICE, count);
    if (result < 0 || *count <= DRM_MAX_PLANES_PER_DEVICE)
        return result;
    for (uint32_t i = 0; i < DRM_MAX_PLANES_PER_DEVICE; i++)
        na_drm_plane_destroy(planes[i]);
    *count = 0;
    return -EINVAL;
}

File: modules/na-std/src/drm/resources.rs
use core::{ffi::CStr, marker::PhantomData, ptr::NonNull};

use crate::{Error, Result, bindings};

const NO_RESOURCE: u32 = u32::MAX;

#[repr(u32)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ConnectorType {
    Unknown = 0,
    Vga = 1,
    DviIntegrated = 2,
    DviDigital = 3,
    DviAnalog = 4,
    Composite = 5,
    SVideo = 6,
    Lvds = 7,
    Component = 8,
    Din9 = 9,
    DisplayPort = 10,
    HdmiA = 11,
    HdmiB = 12,
    Tv = 13,
    EmbeddedDisplayPort = 14,
    Virtual = 15,
    Dsi = 16,
    Dpi = 17,
    Writeback = 18,
}

#[repr(u32)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Connection {
    Connected = 1,
    Disconnected = 2,
    Unknown = 3,
}

#[repr(u32)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum EncoderType {
    None = 0,
    Dac = 1,
    Tmds = 2,
    Lvds = 3,
    TvDac = 4,
    Virtual = 5,
    Dsi = 6,
    DisplayPortMst = 7,
    Dpi = 8,
}

#[repr(u32)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PlaneType {
    Overlay = 0,
    Primary = 1,
    Cursor = 2,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Mode {
    pub clock: u32,
    pub hdisplay: u16,
    pub hsync_start: u16,
    pub hsync_end: u16,
    pub htotal: u16,
    pub hskew: u16,
    pub vdisplay: u16,
    pub vsync_start: u16,
    pub vsync_end: u16,
    pub vtotal: u16,
    pub vscan: u16,
    pub vrefresh: u32,
    pub flags: u32,
    pub mode_type: u32,
    name: [core::ffi::c_char; 32],
}

impl Mode {
    pub const fn new(width: u16, height: u16, refresh: u32) -> Self {
        Self {
            clock: 0,
            hdisplay: width,
            hsync_start: width,
            hsync_end: width,
            htotal: width,
            hskew: 0,
            vdisplay: height,
            vsync_start: height,
            vsync_end: height,
            vtotal: height,
            vscan: 0,
            vrefresh: refresh,
            flags: 0,
            mode_type: 0,
            name: [0; 32],
        }
    }

    pub fn with_name(mut self, name: &CStr) -> Result<Self> {
        let bytes = name.to_bytes();
        if bytes.len() >= self.name.len() {
            return Err(Error::InvalidArgument);
        }
        self.name = [0; 32];
        for (destination, source) in self.name.iter_mut().zip(bytes) {
            *destination = *source as core::ffi::c_char;
        }
        Ok(self)
    }

    pub fn name(&self) -> Option<&CStr> {
        (self.name[0] != 0).then(|| unsafe { CStr::from_ptr(self.name.as_ptr()) })
    }

    pub(crate) fn from_raw(raw: &bindings::DrmModeInfo) -> Self {
        let mut name = raw.name;
        name[31] = 0;
        Self {
            clock: raw.clock,
            hdisplay: raw.hdisplay,
            hsync_start: raw.hsync_start,
            hsync_end: raw.hsync_end,
            htotal: raw.htotal,
            hskew: raw.hskew,
            vdisplay: raw.vdisplay,
            vsync_start: raw.vsync_start,
            vsync_end: raw.vsync_end,
            vtotal: raw.vtotal,
            vscan: raw.vscan,
            vrefresh: raw.vrefresh,
            flags: raw.flags,
            mode_type: raw.mode_type,
            name,
        }
    }

    fn raw(self) -> bindings::DrmModeInfo {
        bindings::DrmModeInfo {
            clock: self.clock,
            hdisplay: self.hdisplay,
            hsync_start: self.hsync_start,
            hsync_end: self.hsync_end,
            htotal: self.htotal,
            hskew: self.hskew,
            vdisplay: self.vdisplay,
            vsync_start: self.vsync_start,
            vsync_end: self.vsync_end,
            vtotal: self.vtotal,
            vscan: self.vscan,
            vrefresh: self.vrefresh,
            flags: self.flags,
            mode_type: self.mode_type,
            name: self.name,
        }
    }
}

#[derive(Clone, Copy)]
pub struct Connector<'a> {
    pub connector_type: ConnectorType,
    pub connection: Connection,
    pub encoder_index: Option<u32>,
    pub crtc_index: Option<u32>,
    pub mm_width: u32,
    pub mm_height: u32,
    pub subpixel: u32,
    pub modes: &'a [Mode],
}

#[derive(Clone, Copy)]
pub struct Crtc {
    pub x: u32,
    pub y: u32,
    pub width: u32,
    pub height: u32,
    pub gamma_size: u32,
    pub mode: Option<Mode>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Encoder {
    pub encoder_type: EncoderType,
    pub crtc_index: Option<u32>,
    pub possible_crtcs: u32,
    pub possible_clones: u32,
}

#[derive(Clone, Copy)]
pub struct Plane<'a> {
    pub crtc_index: Option<u32>,
    pub possible_crtcs: u32,
    pub gamma_size: u32,
    pub plane_type: PlaneType,
    pub formats: &'a [u32],
}

trait RawResource {
    unsafe fn destroy(raw: *mut Self);
}

impl RawResource for bindings::DrmConnector {
    unsafe fn destroy(raw: *mut Self) {
        unsafe { bindings::na_drm_connector_destroy(raw) };
    }
}

impl RawResource for bindings::DrmCrtc {
    unsafe fn destroy(raw: *mut Self) {
        unsafe { bindings::na_drm_crtc_destroy(raw) };
    }
}

impl RawResource for bindings::DrmEncoder {
    unsafe fn destroy(raw: *mut Self) {
        unsafe { bindings::na_drm_encoder_destroy(raw) };
    }
}

impl RawResource for bindings::DrmPlane {
    unsafe fn destroy(raw: *mut Self) {
        unsafe { bindings::na_drm_plane_destroy(raw) };
    }
}

struct ResourceList<'a, T: RawResource> {
    output: NonNull<*mut T>,
    length: u32,
    capacity: u32,
    committed: bool,
    lifetime: PhantomData<&'a mut [*mut T]>,
}

impl<'a, T: RawResource> ResourceList<'a, T> {
    unsafe fn from_raw(output: *mut *mut T, capacity: u32) -> Option<Self> {
        NonNull::new(output).map(|output| Self {
            output,
            length: 0,
            capacity,
            committed: false,
            lifetime: PhantomData,
        })
    }

    fn push(&mut self, raw: NonNull<T>) -> Result<()> {
        if self.length == self.capacity {
            unsafe { T::destroy(raw.as_ptr()) };
            return Err(Error::NoSpace);
        }
        unsafe {
            self.output
                .as_ptr()
                .add(self.length as usize)
                .write(raw.as_ptr())
        };
        self.length += 1;
        Ok(())
    }

    unsafe fn finish(mut self, count: *mut u32) -> i32 {
        let Some(count) = NonNull::new(count) else {
            return Error::InvalidArgument.status();
        };
        unsafe { count.as_ptr().write(self.length) };
        self.committed = true;
        0
    }
}

impl<T: RawResource> Drop for ResourceList<'_, T> {
    fn drop(&mut self) {
        if self.committed {
            return;
        }
        for index in 0..self.length {
            let raw = unsafe { self.output.as_ptr().add(index as usize).read() };
            unsafe { T::destroy(raw) };
        }
    }
}

pub struct ConnectorList<'a>(ResourceList<'a, bindings::DrmConnector>);

impl ConnectorList<'_> {
    pub fn push(&mut self, connector: Connector<'_>) -> Result<()> {
        let info = bindings::DrmConnectorInfo {
            connector_type: connector.connector_type as u32,
            connection: connector.connection as u32,
            encoder_index: connector.encoder_index.unwrap_or(NO_RESOURCE),
            crtc_index: connector.crtc_index.unwrap_or(NO_RESOURCE),
            mm_width: connector.mm_width,
            mm_height: connector.mm_height,
            subpixel: connector.subpixel,
        };
        let raw = NonNull::new(unsafe { bindings::na_drm_connector_create(&info) })
            .ok_or(Error::OutOfMemory)?;
        for mode in connector.modes {
            let status = unsafe { bindings::na_drm_connector_add_mode(raw.as_ptr(), &mode.raw()) };
            if let Err(error) = Error::from_status(status) {
                unsafe { bindings::na_drm_connector_destroy(raw.as_ptr()) };
                return Err(error);
            }
        }
        self.0.push(raw)
    }

    pub(crate) unsafe fn from_raw(
        output: *mut *mut bindings::DrmConnector,
        capacity: u32,
    ) -> Option<Self> {
        unsafe { ResourceList::from_raw(output, capacity) }.map(Self)
    }

    pub(crate) unsafe fn finish(self, count: *mut u32) -> i32 {
        unsafe { self.0.finish(count) }
    }
}

pub struct CrtcList<'a>(ResourceList<'a, bindings::DrmCrtc>);

impl CrtcList<'_> {
    pub fn push(&mut self, crtc: Crtc) -> Result<()> {
        let info = bindings::DrmCrtcInfo {
            x: crtc.x,
            y: crtc.y,
            width: crtc.width,
            height: crtc.height,
            gamma_size: crtc.gamma_size,
            mode_valid: crtc.mode.is_some(),
            mode_info: crtc.mode.map_or_else(Default::default, Mode::raw),
        };
        let raw = NonNull::new(unsafe { bindings::na_drm_crtc_create(&info) })
            .ok_or(Error::OutOfMemory)?;
        self.0.push(raw)
    }

    pub(crate) unsafe fn from_raw(
        output: *mut *mut bindings::DrmCrtc,
        capacity: u32,
    ) -> Option<Self> {
        unsafe { ResourceList::from_raw(output, capacity) }.map(Self)
    }

    pub(crate) unsafe fn finish(self, count: *mut u32) -> i32 {
        unsafe { self.0.finish(count) }
    }
}

pub struct EncoderList<'a>(ResourceList<'a, bindings::DrmEncoder>);

impl EncoderList<'_> {
    pub fn push(&mut self, encoder: Encoder) -> Result<()> {
        let info = bindings::DrmEncoderInfo {
            encoder_type: encoder.encoder_type as u32,
            crtc_index: encoder.crtc_index.unwrap_or(NO_RESOURCE),
            possible_crtcs: encoder.possible_crtcs,
            possible_clones: encoder.possible_clones,
        };
        let raw = NonNull::new(unsafe { bindings::na_drm_encoder_create(&info) })
            .ok_or(Error::OutOfMemory)?;
        self.0.push(raw)
    }

    pub(crate) unsafe fn from_raw(
        output: *mut *mut bindings::DrmEncoder,
        capacity: u32,
    ) -> Option<Self> {
        unsafe { ResourceList::from_raw(output, capacity) }.map(Self)
    }

    pub(crate) unsafe fn finish(self, count: *mut u32) -> i32 {
        unsafe { self.0.finish(count) }
    }
}

pub struct PlaneList<'a> {
    resources: ResourceList<'a, bindings::DrmPlane>,
    device: NonNull<bindings::DrmDevice>,
}

impl PlaneList<'_> {
    pub fn push(&mut self, plane: Plane<'_>) -> Result<()> {
        let info = bindings::DrmPlaneInfo {
            crtc_index: plane.crtc_index.unwrap_or(NO_RESOURCE),
            possible_crtcs: plane.possible_crtcs,
            gamma_size: plane.gamma_size,
            plane_type: plane.plane_type as u32,
        };
        let raw =
            NonNull::new(unsafe { bindings::na_drm_plane_create(self.device.as_ptr(), &info) })
                .ok_or(Error::OutOfMemory)?;
        for format in plane.formats {
            let status = unsafe { bindings::na_drm_plane_add_format(raw.as_ptr(), *format) };
            if let Err(error) = Error::from_status(status) {
                unsafe { bindings::na_drm_plane_destroy(raw.as_ptr()) };
                return Err(error);
            }
        }
        self.resources.push(raw)
    }

    pub(crate) unsafe fn from_raw(
        device: *mut bindings::DrmDevice,
        output: *mut *mut bindings::DrmPlane,
        capacity: u32,
    ) -> Option<Self> {
        Some(Self {
            resources: unsafe { ResourceList::from_raw(output, capacity) }?,
            device: NonNull::new(device)?,
        })
    }

    pub(crate) unsafe fn finish(self, count: *mut u32) -> i32 {
        unsafe { self.resources.finish(count) }
    }
}
